stmhal/timer.c

```diff
@@ -1095,9 +1104,12 @@ STATIC mp_obj_t pyb_timer_callback(mp_obj_t self_in, mp_obj_t callback) {
         __HAL_TIM_DISABLE_IT(&self->tim, TIM_IT_UPDATE);
         self->callback = mp_const_none;
     } else if (mp_obj_is_callable(callback)) {
+        __HAL_TIM_DISABLE_IT(&self->tim, TIM_IT_UPDATE);
```

Comment:
Here's a concrete example which demonstrates the problem. This is the python code:
```python
import pyb

class Heartbeat(object):

    def __init__(self):
        self.led = pyb.LED(4) # 4 = Blue
        print("About to init Timer")
        tim = pyb.Timer(4, freq=1, callback=self.heartbeat_cb)
        print("After init Timer")

    def heartbeat_cb(self, tim):
        print("heartbeat_cb")
        self.led.toggle()

Heartbeat()
```

I also added 3 print statements in pyb_timer_callback:
```C
    } else if (mp_obj_is_callable(callback)) {
        self->callback = callback;
        // start timer, so that it interrupts on overflow
        printf("About to call HAL_TIM_Base_Start_IT\n");
        HAL_TIM_Base_Start_IT(&self->tim);
        printf("After calling HAL_TIM_Base_Start_IT\n");
        HAL_NVIC_EnableIRQ(self->irqn);
        printf("After calling HAL_NVIC_EnableIRQ\n");
    } else {
```

and this is the output. You can see that heartbeat_cb is called immediately as soon as HAL_Base_Start_IT is called evn though clearly a second hasn't passed since initializing the timer.
```
>>> import timer
About to init Timer
About to call HAL_TIM_Base_Start_IT
After calling HAL_TIM_Base_Start_IT
heartbeat_cb
After calling HAL_NVIC_EnableIRQ
After init Timer
>>> heartbeat_cb
heartbeat_cb
heartbeat_cb
heartbeat_cb
```

The above was done without this patch. With this patch, I get this output instead (and the first callback happens about a second after initializing the timer):
```
>>> import timer
About to init Timer
About to call HAL_TIM_Base_Start_IT
After calling HAL_TIM_Base_Start_IT
After calling HAL_NVIC_EnableIRQ
After init Timer
>>> heartbeat_cb
heartbeat_cb
heartbeat_cb
heartbeat_cb
```